kv_storage.h

```diff
@@ -0,0 +1,383 @@
+};
+
+template <typename ENTRY>
+struct ExtractNonThrowingGet<ENTRY, false> {
+  constexpr static bool value = false;  // Default: throw if key not found.
+};
+
+template <typename ENTRY>
+constexpr auto has_overwrite_on_add(int) -> decltype(std::declval<ENTRY>().allow_overwrite_on_add, bool()) {
+  return true;
+}
+
+template <typename ENTRY>
+constexpr bool has_overwrite_on_add(...) {
+  return false;
+}
+
+template <typename ENTRY, bool>
+struct ExtractOverwriteOnAdd {
+ constexpr static bool value = ENTRY::allow_overwrite_on_add;
+ static_assert(!value || std::is_base_of<Nullable, ENTRY>::value, "Entry type must be derived from Nullable.");
+};
+
+template <typename ENTRY>
+struct ExtractOverwriteOnAdd<ENTRY, false> {
+  constexpr static bool value = false;  // Default: don't overwrite on Add() if key already exists.
+};
+
+template <typename ENTRY>
+struct DefaultPolicy {
+  constexpr static bool allow_nonthrowing_get = ExtractNonThrowingGet<ENTRY, has_nonthrowing_get<ENTRY>(0)>::value;
+  constexpr static bool allow_overwrite_on_add = ExtractOverwriteOnAdd<ENTRY, has_overwrite_on_add<ENTRY>(0)>::value;
+};
+
+// Main storage class.
+template <typename ENTRY, typename POLICY = DefaultPolicy<ENTRY>>
+class API {
+ public:
+  typedef ENTRY T_ENTRY;
+  typedef typename std::remove_reference<decltype(std::declval<T_ENTRY>().key())>::type T_KEY;
+  typedef std::function<void(const T_ENTRY&)> T_ENTRY_CALLBACK;
+  typedef std::function<void(const T_KEY&)> T_KEY_CALLBACK;
+  typedef std::function<void()> T_VOID_CALLBACK;
+  typedef POLICY T_POLICY;
+  typedef SpecificEntryTypeKeyNotFoundException<T_ENTRY> T_KEY_NOT_FOUND_EXCEPTION;
+  typedef SpecificEntryKeyAlreadyExistsException<T_ENTRY> T_KEY_ALREADY_EXISTS_EXCEPTION;
+  typedef SpecificEntryShouldExistException<T_ENTRY> T_ENTRY_SHOULD_EXIST_EXCEPTION;
+
+  API(const std::string& stream_name)
+      : stream_(sherlock::Stream<T_ENTRY>(stream_name)), listener_scope_(stream_.Subscribe(listener_)) {}
+
+  typename sherlock::StreamInstance<T_ENTRY>& UnsafeStream() { return stream_; }
+
+  template <typename F>
+  typename StreamInstanceImpl<T_ENTRY>::template ListenerScope<F> Subscribe(F& listener) {
+    return std::move(stream_.Subscribe(listener));
+  }
+
+  std::future<T_ENTRY> AsyncGet(const T_KEY&& key) {
+    std::promise<T_ENTRY> pr;
+    std::future<T_ENTRY> future = pr.get_future();
+    listener_.mq_.EmplaceMessage(new MQMessageGet(key, std::move(pr)));
+    return future;
+  }
+
+  void AsyncGet(const T_KEY&& key, T_ENTRY_CALLBACK on_found, T_KEY_CALLBACK on_not_found) {
+    listener_.mq_.EmplaceMessage(new MQMessageGet(key, on_found, on_not_found));
+  }
+
+  T_ENTRY Get(const T_KEY&& key) {
+    return AsyncGet(std::forward<const T_KEY>(key)).get();
+  }
+
+  std::future<void> AsyncAdd(const T_ENTRY& entry) {
+    std::promise<void> pr;
+    std::future<void> future = pr.get_future();
+    if (listener_.storage_.data.count(entry.key()) && !T_POLICY::allow_overwrite_on_add) {
```

Comment:
We talked of this guy offline, pls move it into `MQMessageAdd::DoIt()`, and add the stream itself as a `DoIt()` parameter everywhere. Thanks!